allen-cell-animated/nucmorph-colorizer
# feature: Alert banner for missing data

Problem
=======
*Estimated review size: medium, 30 minutes*

Closes #288, implementing a warning banner that appears when data is missing.

Solution
========
Oops, this one got a little bigger than I intended! I was trying to make the alert banner system really robust (which. it maybe didn't need to be.)

- Adds the new `AlertBanner` component, which is a wrapper around Ant's Alert component.
  - It enforces certain styling rules (like the `banner` flag) and modifies some of the styling to match UX designs.
  - `AlertBanner` can show a "Don't show again for this dataset" checkbox, the value of which is reported when the banner is closed.
- Adds the `useAlertBanner` hook, which returns a ReactElement list of banners to render, and a callable method for adding new banners.
  - `useAlertBanner` only allows exactly one banner per message, and handles the logic for the "Don't show again" checkbox.
  - Banners will be reset when the dependencies (usually `dataset`) is reset.

Other changes:
- Alert banner is now shown when a bad URL is loaded. (see below verification steps!)
- Adds underlining to focused links for better accessibility :)

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------

Testing error banner:
1. Open preview link. This should fail to load the dataset, and will show the error banner: https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-321/#/viewer?collection=bad

Testing repeated warning:
1. Open preview link. Start playback. https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-321/#/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fassay-dev%2Fcomputational%2Fdata%2FEMT_deliverable_processing%2FTranscription_factor_cell_lines%2FVersion_1%2Fcolorizer_output%2FEOMES%2Fcollections_all.json&dataset=AICS-122_3D-MG-EMT-no-MG_fms_id_592af19bfc414ae1bf05b81967190944
2. Around frame 26, a warning banner should appear.
3. Close the banner and start playback again from frame 20. At frame 26, the same banner should appear. Playing past 26 again will not cause another banner to appear.
4. Check the "Do not show again" checkbox and close the alert. The alert will not reappear.
5. Switch to another dataset. The banner should disappear.

Screenshots (optional):
-----------------------

**Video talk-through (🔊):**

https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/875d9c1f-1b4c-4406-b0ae-1118f4ffe184

![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/d5737131-6c13-4e57-91a9-19b567cfd6f7)
![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/20197034-6407-4b69-8be3-a769b672d348)


Keyfiles (delete if not relevant):
-----------------------
1. `AlertBanner.tsx`
6. `hooks.tsx`: Contains a hook for queueing up alert banners.
